File: Dart/0002.dart

// Definition for singly-linked list.
class ListNode {
  int val;
  ListNode? next;
  ListNode([this.val = 0, this.next]);
}

class Solution {
  ListNode? addTwoNumbers(ListNode? l1, ListNode? l2) {
    // Initialize a dummy node to build the result list
    ListNode dummyHead = ListNode(0);
    ListNode current = dummyHead;
    int carry = 0;

    // Traverse both lists
    while (l1 != null || l2 != null || carry > 0) {
      // Get the values from the nodes if available
      int val1 = (l1 != null) ? l1.val : 0;
      int val2 = (l2 != null) ? l2.val : 0;
      
      // Compute the sum of values and carry
      int sum = val1 + val2 + carry;
      carry = sum ~/ 10; // Update carry for the next digit
      int digit = sum % 10; // Current digit to store in the node

      // Create a new node with the computed digit
      current.next = ListNode(digit);
      current = current.next!; // Move to the next node

      // Move to the next nodes in l1 and l2 if available
      if (l1 != null) l1 = l1.next;
      if (l2 != null) l2 = l2.next;
    }

    // Return the next of dummyHead which points to the head of the result list
    return dummyHead.next;
  }
}

File: Swift/0002.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        let dummyHead = ListNode(0)
        var current = dummyHead
        var carry = 0
        
        var l1 = l1
        var l2 = l2
        
        while l1 != nil || l2 != nil || carry != 0 {
            let val1 = l1?.val ?? 0
            let val2 = l2?.val ?? 0
            
            let sum = val1 + val2 + carry
            carry = sum / 10
            
            current.next = ListNode(sum % 10)
            current = current.next!
            
            l1 = l1?.next
            l2 = l2?.next
        }
        
        return dummyHead.next
    }
}